2018/2018_File_Generation_Features.md

```diff
@@ -0,0 +1,67 @@
+[`python/configurations/`](https://github.com/cfpb/hmda-test-files/tree/master/2018/python/configurations) directory. The following yaml configuration files can be altered to create HMDA submission test files.  
+
+-[`test_filepaths.yaml`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/configurations/test_filepaths.yaml) holds the filepath names for storing clean files, edit test files, log files, and edit reports as well as the paths to the TS and LAR schemas. 
+
+-[`geography_data.yaml`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/configurations/geographic_data.yaml) holds the filepaths for geographic dependency data such as the tract to CBSA file. The configuration includes the column names for the CBSA file and the columns used for file creation. The file also contains a path to a list of zip codes and a dictionary of FIPS state codes. 
+
+-[`clean_file_config.yaml`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/configurations/clean_file_config.yaml) holds values for the test institution, file length, and file name for generating a clean HMDA submission file for testing. Other variables, such as data ranges can also be set in this file. 
+
+Clean files will pass syntax and validity edits. Edit test files will fail at least one (and sometimes more than one) edit. These files will be used as a double check for the implementation of the HMDA edits for the 2018 HMDA Platform. Creating edit test files is a two step process. The generation of edit test files requires a clean data file to be present. If edit files already exist with the same names as those produced by this script, they will be overwritten.
+
+Edit test files are written to the directory indicated in `test_filepaths.yaml` based on edit type, such as /edits_files/test_files/Bank1/syntax/s300.txt. The file name indicates which edit the file is designed to test. If an edit has multiple conditions, a file will be made for each condition in the format /edits_files/test_files/Bank1/syntax/s301_1.txt.
+
+### Changing Edit Logic and the Classes Used for File Creation 
+The FileGenerator Class relies on four other classes, each with its own role in file creation. 
+
+-`lar_generator` in [`lar_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/lar_generator.py) creates raw TS and LAR rows, the values of which are then manipulated to create clean test files or edit test files. 
+
+-`rules_engine` in [`rules_engine.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains logic for each edit. LAR and TS dataframes are passed into this class, and its methods output edit details that show among its fields the edit name, the status of the edit (failed or passed), the number of rows failed by the edit, if any, and the ULI's or 
```

Comment:
After the first line:

`The rules engine can generate an edit report for a passed file. This report contains the following fields:
- edit name
- status (pass/fail)
- number of rows failed
- ULIs/NULIs of rows that failed (as a list)

The report is created as a dataframe that can be passed to other methods of the class and is also written to CSV.

Reply:
Sure!